File: scripts/fetch_hanzi_data.js
#!/usr/bin/env node
/**
 * 批量获取汉字笔画SVG数据
 * 从 Hanzi Writer CDN 获取每个字的笔画路径数据
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// Hanzi Writer 数据源
const HANZI_WRITER_CDN = "https://cdn.jsdelivr.net/npm/hanzi-writer-data@2.0";
const RAW_DIR = path.join(__dirname, "hanzi_raw");

// 确保目录存在
if (!fs.existsSync(RAW_DIR)) {
    fs.mkdirSync(RAW_DIR, { recursive: true });
}

// 请求配置
const MAX_RETRIES = 3;
const RETRY_DELAY = 1000;
const REQUEST_DELAY = 100;

/**
 * 从 Hanzi Writer CDN 获取单个汉字的笔画数据
 */
async function fetchCharData(char) {
    const url = `${HANZI_WRITER_CDN}/${char}.json`;
    
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
        try {
            const data = await new Promise((resolve, reject) => {
                https.get(url, (res) => {
                    let data = '';
                    res.on('data', (chunk) => data += chunk);
                    res.on('end', () => {
                        if (res.statusCode === 200) {
                            try {
                                resolve(JSON.parse(data));
                            } catch (e) {
                                reject(new Error(`JSON解析失败: ${e.message}`));
                            }
                        } else if (res.statusCode === 404) {
                            resolve(null);
                        } else {
                            reject(new Error(`HTTP ${res.statusCode}`));
                        }
                    });
                }).on('error', reject);
            });
            
            return data;
        } catch (error) {
            console.log(`  [重试 ${attempt + 1}] ${char} - ${error.message}`);
            if (attempt < MAX_RETRIES - 1) {
                await sleep(RETRY_DELAY);
            }
        }
    }
    
    console.log(`  [失败] ${char} - 超过最大重试次数`);
    return null;
}

/**
 * 延迟函数
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * 批量获取汉字笔画数据
 */
async function fetchBatch(chars) {
    const results = {};
    const total = chars.length;
    
    for (let i = 0; i < total; i++) {
        const char = chars[i];
        process.stdout.write(`[${i + 1}/${total}] 获取: ${char}`);
        
        // 检查是否已缓存
        const cacheFile = path.join(RAW_DIR, `${char}.raw.json`);
        if (fs.existsSync(cacheFile)) {
            try {
                const data = JSON.parse(fs.readFileSync(cacheFile, 'utf-8'));
                results[char] = data;
                console.log(" (缓存)");
                continue;
            } catch (e) {
                // 缓存损坏，重新获取
            }
        }
        
        const data = await fetchCharData(char);
        if (data) {
            results[char] = data;
            // 保存到缓存
            fs.writeFileSync(cacheFile, JSON.stringify(data, null, 2), 'utf-8');
            console.log(" ✓");
        } else {
            console.log(" ✗");
        }
        
        // 请求间隔
        if (i < total - 1) {
            await sleep(REQUEST_DELAY);
        }
    }
    
    return results;
}

/**
 * 加载已有的笔画数据
 */
function loadExistingData() {
    const results = {};
    if (fs.existsSync(RAW_DIR)) {
        const files = fs.readdirSync(RAW_DIR).filter(f => f.endsWith('.raw.json'));
        for (const file of files) {
            try {
                const data = JSON.parse(fs.readFileSync(path.join(RAW_DIR, file), 'utf-8'));
                const char = file.replace('.raw.json', '');
                results[char] = data;
            } catch (e) {
                // 忽略损坏的文件
            }
        }
    }
    return results;
}

/**
 * 保存合并后的数据到单个文件
 */
function saveMergedData(data) {
    const outputFile = path.join(RAW_DIR, "all_chars_raw.json");
    fs.writeFileSync(outputFile, JSON.stringify(data, null, 2), 'utf-8');
    console.log(`\n已保存到: ${outputFile}`);
    console.log(`总计: ${Object.keys(data).length} 个汉字`);
}

/**
 * 主函数
 */
async function main() {
    console.log("=".repeat(50));
    console.log("批量获取汉字笔画数据");
    console.log("=".repeat(50));
    
    // 导入生字清单
    const { getAllChars, getGradeStats } = require('./primary_school_chars');
    
    // 获取所有生字
    const chars = getAllChars();
    console.log(`小学1-6年级生字: ${chars.length} 个`);
    console.log("-".repeat(40));
    
    // 加载已有数据
    const existing = loadExistingData();
    console.log(`已有数据: ${Object.keys(existing).length} 个`);
    console.log("-".repeat(40));
    
    // 计算需要获取的字
    const toFetch = chars.filter(c => !existing[c]);
    console.log(`需要获取: ${toFetch.length} 个`);
    console.log("-".repeat(40));
    
    let allData;
    if (toFetch.length > 0) {
        // 批量获取
        const newData = await fetchBatch(toFetch);
        
        // 合并数据
        allData = { ...existing, ...newData };
    } else {
        allData = existing;
    }
    
    // 保存合并数据
    saveMergedData(allData);
    
    // 统计
    console.log("-".repeat(40));
    const gradeStats = getGradeStats();
    for (const [grade, count] of Object.entries(gradeStats)) {
        console.log(`  ${grade}: ${count}字`);
    }
}

// 运行
main().catch(console.error);

File: scripts/fetch_hanzi_data.py
# -*- coding: utf-8 -*-
"""
批量获取汉字笔画SVG数据
从 Hanzi Writer CDN 获取每个字的笔画路径数据
"""
import json
import time
import requests
from pathlib import Path

# Hanzi Writer 数据源
HANZI_WRITER_CDN = "https://cdn.jsdelivr.net/npm/hanzi-writer-data@2.0"
RAW_DIR = Path(__file__).parent / "hanzi_raw"
RAW_DIR.mkdir(exist_ok=True)

# 请求头
HEADERS = {
    "User-Agent": "CnStrokesDict/1.0 (Educational App)"
}

# 重试配置
MAX_RETRIES = 3
RETRY_DELAY = 1.0


def fetch_char_data(char: str) -> dict | None:
    """
    从 Hanzi Writer CDN 获取单个汉字的笔画数据
    
    返回格式：
    {
        "char": "好",
        "strokes": ["M 123.45 67.89...", ...],  # SVG路径数据
        "medians": [[...], ...],  # 笔画中心线
        "radStrokes": [0, 1, 2],  # 部首笔画索引
        ...
    }
    """
    url = f"{HANZI_WRITER_CDN}/{char}.json"
    
    for attempt in range(MAX_RETRIES):
        try:
            resp = requests.get(url, headers=HEADERS, timeout=10)
            if resp.status_code == 200:
                return resp.json()
            elif resp.status_code == 404:
                print(f"  [跳过] {char} - 数据不存在")
                return None
            else:
                print(f"  [重试] {char} - HTTP {resp.status_code}")
                time.sleep(RETRY_DELAY)
        except requests.RequestException as e:
            print(f"  [重试] {char} - {e}")
            time.sleep(RETRY_DELAY)
    
    print(f"  [失败] {char} - 超过最大重试次数")
    return None


def fetch_batch(chars: list[str], delay: float = 0.1) -> dict[str, dict]:
    """
    批量获取汉字笔画数据
    
    参数：
        chars: 汉字列表
        delay: 请求间隔（秒），避免被封禁
    
    返回：
        {char: data_dict, ...}
    """
    results = {}
    total = len(chars)
    
    for i, char in enumerate(chars, 1):
        print(f"[{i}/{total}] 获取: {char}", end="")
        
        # 检查是否已缓存
        cache_file = RAW_DIR / f"{char}.raw.json"
        if cache_file.exists():
            try:
                with open(cache_file, "r", encoding="utf-8") as f:
                    results[char] = json.load(f)
                print(" (缓存)")
                continue
            except:
                pass
        
        data = fetch_char_data(char)
        if data:
            results[char] = data
            # 保存到缓存
            with open(cache_file, "w", encoding="utf-8") as f:
                json.dump(data, f, ensure_ascii=False, indent=2)
            print(" ✓")
        else:
            print(" ✗")
        
        # 请求间隔
        if i < total:
            time.sleep(delay)
    
    return results


def load_existing_data() -> dict[str, dict]:
    """加载已有的笔画数据"""
    results = {}
    if RAW_DIR.exists():
        for f in RAW_DIR.glob("*.raw.json"):
            try:
                with open(f, "r", encoding="utf-8") as fp:
                    data = json.load(fp)
                    char = f.stem.replace(".raw", "")
                    results[char] = data
            except:
                pass
    return results


def merge_data(existing: dict, new_data: dict) -> dict:
    """合并已有数据和新数据"""
    merged = existing.copy()
    merged.update(new_data)
    return merged


def save_merged_data(data: dict) -> None:
    """保存合并后的数据到单个文件"""
    output_file = RAW_DIR / "all_chars_raw.json"
    with open(output_file, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    print(f"\n已保存到: {output_file}")
    print(f"总计: {len(data)} 个汉字")


if __name__ == "__main__":
    import sys
    sys.path.insert(0, str(Path(__file__).parent))
    from primary_school_chars import get_all_chars, get_grade_stats
    
    # 获取所有生字
    chars = get_all_chars()
    print(f"小学1-6年级生字: {len(chars)} 个")
    print("-" * 40)
    
    # 加载已有数据
    existing = load_existing_data()
    print(f"已有数据: {len(existing)} 个")
    print("-" * 40)
    
    # 计算需要获取的字
    to_fetch = [c for c in chars if c not in existing]
    print(f"需要获取: {len(to_fetch)} 个")
    print("-" * 40)
    
    if to_fetch:
        # 批量获取
        new_data = fetch_batch(to_fetch)
        
        # 合并数据
        all_data = merge_data(existing, new_data)
    else:
        all_data = existing
    
    # 保存合并数据
    save_merged_data(all_data)
    
    # 统计
    print("-" * 40)
    grade_stats = get_grade_stats()
    for grade, count in grade_stats.items():
        print(f"  {grade}: {count}字")
